### PR TITLE
feat(eventProcessor): Adds EventProcessor and BatchEventProcessor

### PR DESCRIPTION
**Pull request by mfahadahmed**
## Summary
- Introduces an EventProcessor interface.
- Introduces a BatchEventProcessor
- Renamed EventAttributes class file name to EventTags

Buffering events within a queue before dispatching is an optimization that should prevent SDK implementations from exhausting resources while increasing throughput. This implementation relies on a `BlockingCollection` to buffer events received from one-to-many producers. A single consumer thread continuously polls from this queue to build a batch before emitting the batched `LogEvent`.

## Test Plan
- Added unit tests

### CODE REVIEW

File: OptimizelySDK.Tests/EventTests/BatchEventProcessorTest.cs [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+        {
+            var countdownEvent = new CountdownEvent(1);
+            var eventDispatcher = new TestEventDispatcher(countdownEvent);
+            SetEventProcessor(eventDispatcher);
+
+            for (int i = 0; i < MAX_BATCH_SIZE; i++)
+            {
+                UserEvent userEvent = BuildConversionEvent(EventName);
+                EventProcessor.Process(userEvent);
+                eventDispatcher.ExpectConversion(EventName, TestUserId);
```
Comment (mikeng13) [inline OptimizelySDK.Tests/EventTests/BatchEventProcessorTest.cs]: Hmm, so with the batch processor, we expect the userEvents to be grouped up to the `MAX_BATCH_SIZE` and then only one `LogEvent` should make it to the eventDispatcher on flush

File: OptimizelySDK.Tests/EventTests/CanonicalEvent.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            ExperimentId = experimentId;
+            VariationId = variationId;
+            EventName = eventName;
+            VisitorId = visitorId;
+            Attributes = attributes ?? new Dictionary<string, object>();
+            Tags = tags ?? new EventTags();
+        }
+
+        // if u really need to use Equals, use hashcode as well.
+        // i don't thiknk so it's need, u may use overloaded operator to check if it's equal or not.
```
Comment (mikeng13) [inline OptimizelySDK.Tests/EventTests/CanonicalEvent.cs]: lol this comment, can we fix it up please

File: OptimizelySDK.Tests/EventTests/CanonicalEvent.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
+﻿using OptimizelySDK.Entity;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace OptimizelySDK.Tests.EventTests
+{
+    public class CanonicalEvent
```
Comment (mikeng13) [inline OptimizelySDK.Tests/EventTests/CanonicalEvent.cs]: Why are we calling this a `CanonicalEvent`?

File: OptimizelySDK/Event/BatchEventProcessor.cs [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+                FlushQueue();
+            }
+        }
+
+        private bool ShouldSplit(UserEvent userEvent) {
+            if (CurrentBatch.Count == 0) {
+                return false;
+            }
+
+            EventContext currentContext = CurrentBatch.Last().Context;
```
Comment (mikeng13) [inline OptimizelySDK/Event/BatchEventProcessor.cs]: Do we need to lock here since we are reading from the CurrentBatch?

File: OptimizelySDK/Event/BatchEventProcessor.cs [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+            lock (mutex)
+            {
+                toProcessBatch = new List<UserEvent>(CurrentBatch);
+                CurrentBatch.Clear();
+            }
+            
+
+            LogEvent logEvent = EventFactory.CreateLogEvent(toProcessBatch.ToArray(), Logger);
+
+            NotificationCenter?.SendNotifications(NotificationCenter.NotificationType.LogEvent, logEvent);
```
Comment (mikeng13) [inline OptimizelySDK/Event/BatchEventProcessor.cs]: Please split up the notifications into another PR so we can review them independently.

File: OptimizelySDK.Tests/EventTests/CanonicalEvent.cs (hunk 3 above)
Comment (mfahadahmed) [reply OptimizelySDK.Tests/EventTests/CanonicalEvent.cs]: The name is same as in the Java SDK:
https://github.com/optimizely/java-sdk/blob/88d0e6a33ce26b30a01765c57e274c2f86d2dc17/core-api/src/test/java/com/optimizely/ab/EventHandlerRule.java#L161

File: OptimizelySDK.Tests/EventTests/BatchEventProcessorTest.cs (hunk 1 above)
Comment (mfahadahmed) [reply OptimizelySDK.Tests/EventTests/BatchEventProcessorTest.cs]: Yes, I also added logger verification in the unit test to validate that DispatchEvent gets called only once.

File: OptimizelySDK/Event/BatchEventProcessor.cs (hunk 4 above)
Comment (mfahadahmed) [reply OptimizelySDK/Event/BatchEventProcessor.cs]: Added lock statement.

File: OptimizelySDK.Tests/EventTests/CanonicalEvent.cs (hunk 2 above)
Comment (mfahadahmed) [reply OptimizelySDK.Tests/EventTests/CanonicalEvent.cs]: Done :)

File: OptimizelySDK/Utils/DateTimeExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using System;
+
+namespace OptimizelySDK.Utils
+{
+    public static class DateTimeExtensions
```
Comment (mnoman09) [inline OptimizelySDK/Utils/DateTimeExtensions.cs]: We already have [DateTimeUtils ](https://github.com/optimizely/csharp-sdk/blob/master/OptimizelySDK/Utils/DateTimeUtils.cs) cIass, I think we can use that instead of adding one more Datetime util class. So we can either remove that or we can move these functions into that class.

File: OptimizelySDK/Event/BatchEventProcessor.cs [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+
+namespace OptimizelySDK.Event
+{
+    /**
+     * BatchEventProcessor is a batched implementation of the {@link EventProcessor}
+     *
+     * Events passed to the BatchEventProcessor are immediately added to a BlockingQueue.
+     *
+     * The BatchEventProcessor maintains a single consumer thread that pulls events off of
+     * the BlockingQueue and buffers them for either a configured batch size or for a
```
[the remark this answers is not in the document]
Comment (msohailhussain) [reply OptimizelySDK/Event/BatchEventProcessor.cs]: No queue is fine.

File: OptimizelySDK.Tests/EventTests/CanonicalEvent.cs [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        private Dictionary<string, object> Attributes;
+        private EventTags Tags;
+
+        public CanonicalEvent(string experimentId, string variationId, string eventName, string visitorId, Dictionary<string, object> attributes, EventTags tags)
+        {
+            ExperimentId = experimentId;
+            VariationId = variationId;
+            EventName = eventName;
+            VisitorId = visitorId;
+            Attributes = attributes ?? new Dictionary<string, object>();
```
Comment (aliabbasrizvi) [inline OptimizelySDK.Tests/EventTests/CanonicalEvent.cs]: Wouldn't it be better to make this `new UserAttributes()`

File: OptimizelySDK.Tests/EventTests/CanonicalEvent.cs [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+            Tags = tags ?? new EventTags();
```
Comment (aliabbasrizvi) [inline OptimizelySDK.Tests/EventTests/CanonicalEvent.cs]: I feel the file for EventTags should be called `EventTags` and not `EventAttributes`.

File: OptimizelySDK.Tests/EventTests/CanonicalEvent.cs (hunk 8 above)
Comment (aliabbasrizvi) [reply OptimizelySDK.Tests/EventTests/CanonicalEvent.cs]: Feedback still not addressed.

File: OptimizelySDK.Tests/EventTests/TestEventDispatcher.cs [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+                return;
+
+            foreach (var visitor in visitors)
+            {
+                foreach (var snapshot in visitor.Snapshots)
+                {
+                    var decisions = snapshot.Decisions ?? new Decision[1] { new Decision() };
+                    foreach (var decision in decisions)
+                    {
+                        foreach (var eevent in snapshot.Events)
```
Comment (aliabbasrizvi) [inline OptimizelySDK.Tests/EventTests/TestEventDispatcher.cs]: I am assuming `event` is a keyword?

File: OptimizelySDK/Event/BatchEventProcessor.cs [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+     * maximum duration before the resulting LogEvent is sent to the NotificationManager.
+     */
+    public class BatchEventProcessor: EventProcessor, IDisposable
+    {
+        private const int DEFAULT_BATCH_SIZE = 10;
+        private static readonly TimeSpan DEFAULT_FLUSH_INTERVAL = TimeSpan.FromMinutes(1);
```
Comment (aliabbasrizvi) [inline OptimizelySDK/Event/BatchEventProcessor.cs]: I think I previously mentioned a default of 30 seconds. We should change it to that to be consistent.

File: OptimizelySDK.Tests/EventTests/TestEventDispatcher.cs (hunk 10 above)
Comment (msohailhussain) [reply OptimizelySDK.Tests/EventTests/TestEventDispatcher.cs]: yeah but u can use `@` with the reserved keywords to use as variable.
